pi5/pinctrl_utils.go

```diff
@@ -63,6 +74,34 @@ func (b *pinctrlpi5) findPathFromAlias(nodeName string) (string, error) {
 	return nodePath, err
 }
 
+// read in 'numCells' 32 bit chunks from byteContents, the bytestream outputted from reading the file '/ranges'. Convert bytes into their uint64 value
+func parseCells(numCells uint32, byteContents *[]byte) (uint64, error) {
+	var parsedValue uint64
+
+	if len(*byteContents) < int(numCells)*4 {
+		errorMsg := "num cells was: " + string(numCells) + ", but there aren't enough bytes to read in from inputted bytestream"
+		return 0, errors.New(errorMsg)
+	}
+
+	switch numCells {
+
+	// reading in 32 bits. regardless we must convert to a 64 bit address so we add a bunch of 0s to the beginning.
+	case 1:
+		parsedValue = uint64(binary.BigEndian.Uint32((*byteContents)[:(4 * numCells)]))
+
+	// reading in 64 bits
+	case 2:
+		parsedValue = binary.BigEndian.Uint64((*byteContents)[:(4 * numCells)])
+
+	// reading in more than 64 bits. we only want the last 64 bits of the address so we cut off the other portion
+	default:
+		parsedValue = binary.BigEndian.Uint64((*byteContents)[(4 * (numCells - 2)):(4 * numCells)])
```

Comment:
1 cell is 32 bits. 
Since we only want the last 64 its of the address, we only want the last 2 cells from the byte contents, regardless of how long it is. 
(this is why "numCells - 2" is written)

The output from reading in any file in the device tree is in bytes. However, we interpret the device tree using units of cells, which are 32 bit chunks of data.
1 cell = 32 bits = 4bytes 
(this is why every cell is multiplied by 4).

Reply:
These cells are entirely different than the pinData information, which comes after we've parsed the device tree and located our gpio chip in memory.